src/Pos/Update/Poll/Pure.hs

```diff
@@ -0,0 +1,123 @@
+import           Pos.Crypto                (hash)
+import           Pos.Types                 (SoftwareVersion (..))
+import           Pos.Update.Core           (UpdateProposal (..))
+import           Pos.Update.Poll.Class     (MonadPoll (..), MonadPollRead (..))
+import qualified Pos.Update.Poll.PollState as Poll
+import           Pos.Update.Poll.Types     (BlockVersionState (..),
+                                            DecidedProposalState (..),
+                                            UndecidedProposalState (..),
+                                            cpsSoftwareVersion, propStateToEither,
+                                            psProposal)
+
+newtype PurePoll a = PurePoll
+    { getPurePoll :: NamedPureLogger (State Poll.PollState) a
+    } deriving (Functor, Applicative, Monad, CanLog, HasLoggerName)
+
+runPurePollWithLogger :: Poll.PollState -> PurePoll a -> (a, Poll.PollState, [LogEvent])
+runPurePollWithLogger ps =
+    (\((a, c), b) -> (a, b, c)) . flip runState ps . runNamedPureLog . getPurePoll
+
+evalPurePollWithLogger :: Poll.PollState -> PurePoll a -> a
+evalPurePollWithLogger r = view _1 . runPurePollWithLogger r
+
+execPurePollWithLogger :: Poll.PollState -> PurePoll a -> Poll.PollState
+execPurePollWithLogger r = view _2 . runPurePollWithLogger r
+
+instance MonadPollRead PurePoll where
+    getBVState bv = PurePoll $ use $ Poll.psBlockVersions . at bv
+    getProposedBVs = PurePoll $ use $ Poll.psBlockVersions . to HM.keys
+    getEpochProposers = PurePoll $ use $ Poll.psEpochProposers
+    getCompetingBVStates = PurePoll $ use $ Poll.psBlockVersions . to HM.toList
+    getAdoptedBVFull = PurePoll $ use $ Poll.psAdoptedBV
+    getLastConfirmedSV an = PurePoll $ use $ Poll.psConfirmedANs . at an
+    getProposal ui = PurePoll $ use $ Poll.psActiveProposals . at ui
+    getProposalsByApp an = PurePoll $ do
+        activeProposalsIndices <- use Poll.psActivePropsIdx
+        activeProposals        <- use Poll.psActiveProposals
+        propGetByApp activeProposalsIndices activeProposals
+      where
+        propGetByApp appHashmap upIdHashmap =
+            case HM.lookup an appHashmap of
+                Nothing -> do
+                    logDebug $
+                        "getProposalsByApp: unknown application name " <> pretty an
+                    pure []
+                Just hashset -> do
+                    let uidList = toList hashset
+                        propStateList = map (\u -> (u, HM.lookup u upIdHashmap)) uidList
+                    fromMaybe [] . mapM snd <$> filterM filterFun propStateList
+        filterFun (uid, Nothing) = do
+            logWarning $ "getProposalsByApp: unknown update id " <> pretty uid
+            pure False
+        filterFun (_, Just _) = pure True
+    getConfirmedProposals = PurePoll $ use $ Poll.psConfirmedProposals . to HM.elems
+    getEpochTotalStake ei =
+        PurePoll $ use $ Poll.psFullRichmenData . to ((Just . fst) <=< HM.lookup ei)
```

Comment:
Perhaps the code around here could be made a little bit less noisy by using `uses` instead of `to`.

Instead of `use (l . to f)` just write `use l f`.